tests/test_highlighting/highlighter.py

```diff
@@ -0,0 +1,157 @@
+            pass # If we wanted text, but we don't get syntactic coloring, then we tolerate
+        else:
+            # in all other cases, an error is returned indicating the number of the problematic character
+            return False,i
+
+    return True,-1
+
+
+# Returns whether a portion of the string has already been " highlighted ".
+def notYetUsed(string):
+    for c in string:
+        if c not in [" ","\n"]:
+            return False
+    return True
+
+
+# This function simulates the operation of Ace on a
+# string provided as input, and "colors" it using the variable Tokencode.
+
+# Be careful, this function only works when the coloring
+# uses only one state in its automaton (`start`)
+
+def simulateRulesWithoutToken(Rules,code):
+    # we initialize the result variable
+    Output = ""
+    for c in code:
+        if c == "\n": Output += "\n"
+        else: Output += " "
+
+    # We go through all the rules 
+    regRule = Rules["start"][0]
+    for reg in regRule:
+
+        # for each rule, we create the regex
+        regComp = re.compile(reg["regex"], re.MULTILINE)
+
+        # In case the rule does not use several groups
+        if type(reg['token']) == str or len(reg['token']) == 1:
+
+            # we get the type of the syntactic coloring
+            if  type(reg['token']) == str : Tokcode = Tokencode[reg['token']]
+            else: Tokcode = Tokencode[reg['token'][0]]
+
+
+            for match in regComp.finditer(code):
+                # For each match, we see if this sequence has already been used
+                if notYetUsed(Output[match.start():match.end()]):
+
+                    start  = match.start()
+                    length = match.end() - start
+
+                    # we replace in the output spaces by the token
+                    Output = Output[:start] + Tokcode * length + Output[start + length:]
+
+
+        else: # In case the rule uses several groups
+
+            for match in regComp.finditer(code):
+                # For each match, we see if this sequence has already been used
+                if notYetUsed(Output[match.start():match.end()]):
+
+                    pos = match.start()
+                    for i,submatch in enumerate(match.groups()):
+                        # For each of the subgroups of the match
+                        # we replace the spaces by the coloring code
+                        tok = reg['token'][i%len(reg['token'])]
+                        Output = Output[:pos] + Tokencode[tok] * len(submatch) + Output[pos + len(submatch):]
+                        pos += len(submatch)
+
+    return Output
+
+
+#######################################################################################
+#######################################################################################
+#######################################################################################
+
+
+
+
+import json
+
+# open data for regex
+file = open('data.json') 
```

Comment:
Data driven tests are great, but JSON is an annoying format to edit. Especially once the examples will span multiple lines.

I don't think that there's a lot of benefit to putting tests in JSON vs. putting them in plain Python. Given a couple of well-chosen helpers, tests in Python could look like this:

```py
class HighlighterTests:
  def test_print_is_recognized_as_keyword(self):
    self.assertHighlighted(
      'print hello world',
      'KKKKK TTTTT TTTTT',
    )
```

That's equally as readable and convenient, but allows for more flexibility in the future (and integrates nicely into existing test runner frameworks)

Reply:
If we have multiline tests we could make Python helpers that do this:


```py
  def test_multiline_example(self):
    self.assertHighlighted(
      'naam is bert',
      'VVVV KK TTTT',

      'print naam er zit een banaan in je oor',
      'KKKKK VVVV TT TTT TTT TTTTTT TT TT TTT',
    )
```

Which will be impossible to achieve in JSON.